interbtc-spec/docs/source/spec/staking.rst

```diff
@@ -0,0 +1,550 @@
+.. _staking:
+
+Staking
+=======
+
+Overview
+~~~~~~~~
+
+This pallet is very similar to the :ref:`rewards` pallet - it is also based on the `Scalable Reward Distribution <https://solmaz.io/2019/02/24/scalable-reward-changing/>`_ algorithm. The reward pallet keeps track of how much rewards vaults have earned. However, when nomination is enabled, there needs to be a way to relay parts of the vault's rewards to its nominators. Furthermore, the nominator's collaterals can be consumed, e.g., when a redeem is cancelled. This pallet is responsible for both tracking the rewards, and the current amount of contributed collaterals of vaults and nominators.
+
+The idea is to have one reward pool per vault, where both the vault and all of its nominators have a stake equal to their contributed collateral. However, when collateral is consumed, either in :ref:`cancelRedeem` or :ref:`liquidateVault`, the collateral of each of these stakeholders should decrease proportionally to their stake. To be able to achieve this without iteration, in addition to tracking ``RewardPerToken``, a similar value ``SlashPerToken`` is introduced. Similarly, in addition to ``RewardTally``, we now also maintain a ``SlashTally`` is for each stakeholder. When calculating a reward for a stakeholder, a compensated stake is calculated, based on ``Stake``, ``SlashPerToken`` and ``SlashTally``. 
+
+When a vault opts out of nomination, all nominators should receive their collateral back. This is achieved by distributing all funds from the vault's shared collateral as rewards. However, a vault is free to opt back into nominator after having opted out. It is possible for the vault to do this before all nominators have withdrawn their reward. To ensure that the bookkeeping remains intact for the nominators to get their rewards at a later point, all variables are additionally indexed by a nonce, which increases every time a vault opts out of nomination. Effectively, this create a new pool for every nominated period.
+
+.. note:: Most of the functions in this pallet that have a ``_at_index`` also have a version without this suffix that does not take a ``nonce`` argument, and instead uses the value stored in :ref:`Nonce`. For brevity, these functions without the suffix are omitted in this specification. 
+
+
+Data Model
+~~~~~~~~~~
+
+Maps
+----
+
+TotalStake
+..........
+
+Maps ``(currencyId, nonce, vaultId)`` to the total stake deposited by the given vault and its nominators, with the given nonce and currencyId.
```

Comment:
I'm contemplating whether we should remove `currencyId`, currently (and possibly also after deciding upon the multi-collateral scheme) we would only have one staking and one reward currency which makes this input useless.

Reply:
It's also a bit confusing since `stake` (i.e. collateral) is stored under the reward currency ID.

Reply:
I would be in favour of removing it, but probably best to wait until the multi collateral scheme is finalized